crates/iota-data-ingestion-core/src/executor/v2.rs

```diff
@@ -0,0 +1,286 @@
+///
+/// struct CustomWorker;
+///
+/// #[async_trait]
+/// impl Worker for CustomWorker {
+///     type Message = ();
+///     type Error = IngestionError;
+///
+///     async fn process_checkpoint(
+///         &self,
+///         checkpoint: Arc<CheckpointData>,
+///     ) -> Result<Self::Message, Self::Error> {
+///         // custom processing logic.
+///         println!(
+///             "Processing Local checkpoint: {}",
+///             checkpoint.checkpoint_summary.to_string()
+///         );
+///         Ok(())
+///     }
+/// }
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let concurrency = 5;
+///     let progress_store = FileProgressStore::new("progress.json").await.unwrap();
+///     let config = CheckpointReaderConfig {
+///         ingestion_path: Some(PathBuf::from("./chk".to_string())),
+///         ..Default::default()
+///     };
+///     let mut executor = IndexerExecutor::new(
+///         progress_store,
+///         1, // number of registered WorkerPools.
+///         DataIngestionMetrics::new(&Registry::new()),
+///         CancellationToken::new(),
+///     );
+///     // register a worker pool with 5 workers to process checkpoints in parallel
+///     let worker_pool = WorkerPool::new(
+///         CustomWorker,
+///         "local_reader".to_string(),
+///         concurrency,
+///         Default::default(),
+///     );
+///     // register the worker pool to the executor.
+///     executor.register(worker_pool).await.unwrap();
+///     // run the ingestion pipeline.
+///     executor.run(config).await.unwrap();
+/// }
+/// ```
+pub struct IndexerExecutor<P> {
+    inner: IndexerExecutorV1<P>,
+}
+
+impl<P: ProgressStore> IndexerExecutor<P> {
+    pub fn new(
+        progress_store: P,
+        number_of_jobs: usize,
+        metrics: DataIngestionMetrics,
+        token: CancellationToken,
+    ) -> Self {
+        Self {
+            inner: IndexerExecutorV1::new(progress_store, number_of_jobs, metrics, token),
+        }
+    }
+
+    /// Registers new worker pool in executor.
+    pub async fn register<W: Worker + 'static>(
+        &mut self,
+        pool: WorkerPool<W>,
+    ) -> IngestionResult<()> {
+        self.inner.register(pool).await
+    }
+
+    pub async fn update_watermark(
+        &mut self,
+        task_name: String,
+        watermark: CheckpointSequenceNumber,
+    ) -> IngestionResult<()> {
+        self.inner.update_watermark(task_name, watermark).await
+    }
+
+    pub async fn read_watermark(
+        &mut self,
+        task_name: String,
+    ) -> IngestionResult<CheckpointSequenceNumber> {
+        self.inner.read_watermark(task_name).await
+    }
+
+    /// Main executor loop.
+    ///
+    /// # Error
+    ///
+    /// Returns an [`IngestionError::EmptyWorkerPool`] if no worker pool was
+    /// registered.
+    pub async fn run(
+        mut self,
+        config: CheckpointReaderConfig,
+    ) -> IngestionResult<ExecutorProgress> {
```

Comment:
It is not clear to me why we need an `executor::v2`. All other methods here just use `inner` and it seems to me that we could just a new method `run_with_config` that takes `CheckpointReaderConfig` that encapsulates the new reader.